jbmc/src/java_bytecode/assignments_from_json.cpp

```diff
@@ -0,0 +1,837 @@
+
+#include "ci_lazy_methods_needed.h"
+#include "java_static_initializers.h"
+#include "java_string_library_preprocess.h"
+#include "java_string_literals.h"
+#include "java_utils.h"
+
+#include <goto-programs/class_identifier.h>
+#include <util/allocate_objects.h>
+#include <util/array_element_from_pointer.h>
+#include <util/expr_initializer.h>
+#include <util/prefix.h>
+#include <util/unicode.h>
+
+/// Values passed around between most functions of the recursive deterministic
+/// assignment algorithm entered from \ref assign_from_json.
+/// The values in a given `det_creation_infot` are never reassigned, but the
+/// ones that are not marked `const` may be mutated.
+struct det_creation_infot
+{
+  /// Code block to append all new code to for the deterministic assignments.
+  code_blockt &block;
+
+  /// Handles allocation of new symbols, adds them to its symbol table (which
+  /// will usually be the same as the `symbol_table` of this struct) and keeps
+  /// track of them so declarations for them can be added by the caller before
+  /// `block`.
+  allocate_objectst &allocate_objects;
+
+  /// Used for looking up symbols corresponding to Java classes and methods.
+  symbol_table_baset &symbol_table;
+
+  /// Where runtime types differ from compile-time types, we need to mark the
+  /// runtime types as needed by lazy methods.
+  optionalt<ci_lazy_methods_neededt> &needed_lazy_methods;
+
+  /// Map to keep track of reference-equal objects. Each entry has an ID (such
+  /// that any two reference-equal objects have the same ID) and the expression
+  /// for the symbol that all these references point to.
+  std::unordered_map<std::string, det_creation_referencet> &references;
+
+  /// Source location associated with the newly added codet.
+  const source_locationt &loc;
+
+  /// Maximum value allowed for any (constant or variable length) arrays in user
+  /// code.
+  size_t max_user_array_length;
+
+  /// Used for the workaround for enums only.
+  /// See assign_enum_from_json.
+  const java_class_typet &declaring_class_type;
+};
+
+static java_class_typet
+followed_class_type(const exprt &expr, const symbol_table_baset &symbol_table)
+{
+  const pointer_typet &pointer_type = to_pointer_type(expr.type());
+  const java_class_typet &java_class_type =
+    to_java_class_type(namespacet{symbol_table}.follow(pointer_type.subtype()));
+  return java_class_type;
+}
+
+static bool
+has_enum_type(const exprt &expr, const symbol_table_baset &symbol_table)
+{
+  return followed_class_type(expr, symbol_table).get_is_enumeration();
+}
+
+/// This function is used as a workaround until reference-equal objects defined
+/// across several classes are tracked correctly. Once reference-equality works
+/// in all cases, this function can be removed.
+/// Until then, in the case of an enum expression that needs to be assigned a
+/// value, we distinguish between two cases:
+/// 1) the enum expression is declared in its own type - in this case,
```

Comment:
Not sure what "declared in its own type" means, could you expand the comment with an example?

Reply:
Rephrased it, added examples, and renamed it to `is_enum_with_type_equal_to_declaring_type` (long name but more descriptive and it's only used in one place anyway; let me know if you have an idea for a better name).